R/data.table.R

```diff
@@ -2887,9 +2887,13 @@ setDT = function(x, keep.rownames=FALSE, key=NULL, check.names=FALSE) {
     rn = if (!identical(keep.rownames, FALSE)) rownames(x) else NULL
     setattr(x, "row.names", .set_row_names(nrow(x)))
     if (check.names) setattr(x, "names", make.names(names(x), unique=TRUE))
+
+    # setalloccol results in a shallow copy. Must be performed before class setting,
+    # to have the class apply only to the new copy. #4784
+    setalloccol(x)
```

Comment:
```suggestion
    # setalloccol results in a shallow copy. Must before performed before setting other attributes. #4784
    setalloccol(x)

    rn = if (!identical(keep.rownames, FALSE)) rownames(x) else NULL
    setattr(x, "row.names", .set_row_names(nrow(x)))
    if (check.names) setattr(x, "names", make.names(names(x), unique=TRUE))
```

Move `setalloccol` up to avoid future corner case with row order or names